vespalib/src/vespa/vespalib/datastore/data_store_explorer.cpp

```diff
@@ -6,10 +6,275 @@
+    object.setLong("allocated_entries", _allocated_entries);
+    object.setLong("used_entries", _used_entries);
+    object.setLong("dead_entries", _dead_entries);
+    object.setLong("hold_entries", _hold_entries);
+    object.setLong("extra_used_bytes", _extra_used_bytes);
+    object.setLong("extra_hold_bytes", _extra_hold_bytes);
+}
+
+class BufferTypeStats {
+    uint32_t _type_id;
+    uint32_t _entry_size;
+    uint32_t _array_size;
+    uint32_t _max_entries;
+    BufferTypeStateStats _active;
+    BufferTypeStateStats _hold;
+
+public:
+    BufferTypeStats() noexcept;
+    bool is_initialized() const noexcept { return _active.buffers() != 0 || _hold.buffers() != 0; }
+    void aggregate(const BufferState& state) noexcept;
+    void stats_to_slime(Inserter& inserter);
+    uint32_t type_id() const noexcept { return _type_id; }
+    uint64_t entry_size() const noexcept { return _entry_size; }
+    const BufferTypeStateStats& active() const noexcept { return _active; }
+    const BufferTypeStateStats& hold() const noexcept { return _hold; }
+    uint32_t buffers() const noexcept { return active().buffers() + hold().buffers(); }
+    uint32_t allocated_entries() const noexcept { return active().allocated_entries() + hold().allocated_entries(); }
+    uint32_t used_entries() const noexcept { return active().used_entries() + hold().used_entries(); }
+    uint32_t dead_entries() const noexcept { return active().dead_entries() + hold().dead_entries(); }
+    uint32_t hold_entries() const noexcept { return active().hold_entries() + hold().hold_entries(); }
+    uint64_t extra_used_bytes() const noexcept { return active().extra_used_bytes() + hold().extra_used_bytes(); }
+    uint64_t extra_hold_bytes() const noexcept { return active().extra_hold_bytes() + hold().extra_hold_bytes(); }
+    uint64_t used_bytes() const noexcept { return used_entries() * entry_size() + extra_used_bytes(); }
+    uint64_t allocated_bytes() const noexcept { return allocated_entries() * entry_size() + extra_used_bytes(); }
+    uint64_t hold_bytes() const noexcept { return hold_entries() * entry_size() + extra_hold_bytes(); }
+    uint64_t dead_bytes() const noexcept { return dead_entries() * entry_size(); }
+};
+
+BufferTypeStats::BufferTypeStats() noexcept
+    : _type_id(0),
+      _entry_size(0),
+      _array_size(0),
+      _max_entries(0),
+      _active(),
+      _hold()
+{
+}
+
+void
+BufferTypeStats::aggregate(const BufferState& state) noexcept
+{
+    if (!is_initialized()) {
+        _type_id = state.getTypeId();
+        auto type_handler = state.getTypeHandler();
+        _entry_size = type_handler->entry_size();
+        _array_size = type_handler->getArraySize();
+        _max_entries = type_handler->get_max_entries();
+    }
+    switch (state.getState()) {
+        case BufferState::State::ACTIVE:
+            _active.aggregate(state);
+        break;
+        case BufferState::State::HOLD:
+            _hold.aggregate(state);
```

Comment:
Even though `State::FREE` is a no-op, should there be a `break` here to avoid fallthrough in the case that this changes?